File: python/04_010.py
class UserService:
  """
  Service for user operations with email notification.
  """
  def __init__(self, user_repository, email_service):
     self.user_repository = user_repository
     self.email_service = email_service
  
  def get_user(self, username):
     return self.user_repository.find_by_username(
        username)
  
  def create_user(self, user):
     if self.user_repository.exists_by_username(
           user.get_username()):
        raise UserAlreadyExistsException()
     self.user_repository.save(user)
     self.email_service.send_welcome_email(
        user.get_email())

File: cpp/04_010.cpp
#include <string>
#include <memory>
#include <stdexcept>

class User; // Forward declaration
class UserRepository; // Forward declaration
class EmailService; // Forward declaration
class UserAlreadyExistsException; // Forward declaration

class UserService {
private:
    std::shared_ptr<UserRepository> userRepository;
    std::shared_ptr<EmailService> emailService;
    
public:
    UserService(std::shared_ptr<UserRepository> repository, 
                std::shared_ptr<EmailService> email) 
        : userRepository(repository), emailService(email) {}
 
    User getUser(const std::string& username) {
        return userRepository->findByUsername(username);
    }
 
    void createUser(const User& user) {
        if (userRepository->existsByUsername(
                user.getUsername())) {
            throw UserAlreadyExistsException();
        }
        userRepository->save(user);
        emailService->sendWelcomeEmail(user.getEmail());
    }
};